plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchangeSink.java

```diff
@@ -0,0 +1,294 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.exchange;
+
+import io.airlift.slice.BasicSliceOutput;
+import io.airlift.slice.Slice;
+import io.airlift.slice.SliceOutput;
+import io.airlift.slice.Slices;
+import io.trino.spi.exchange.ExchangeSink;
+import org.openjdk.jol.info.ClassLayout;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+import javax.crypto.SecretKey;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.net.URI;
+import java.util.ArrayDeque;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Queue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReentrantLock;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Throwables.throwIfUnchecked;
+import static io.airlift.slice.SizeOf.estimatedSizeOf;
+import static java.lang.Math.min;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.CompletableFuture.allOf;
+import static java.util.concurrent.CompletableFuture.completedFuture;
+
+@ThreadSafe
+public class FileSystemExchangeSink
```

Comment:
Somehow it is super difficult for me to follow locking strategy in this class.

The locking strategy and interactions in this class are inherently complex. There are per-partition locks, conditioned `lock` for the buffer pool and the lock protecting the `writersMap`.

It supper difficult for me to give a definite answer whether all interactions (such as "add after abort", "abort after finish", "abort during finish") will result in a correct behavior.

To make it a little simpler it feels like each locking strategy has to be encapsulated in a separate entity.

I would like to kindly suggest a following structure.

First, let's encapsulate buffer pool. This can be done by introducing an inner, thread safe class with a following interface:

```
    @ThreadSafe
    private static class BufferPool
    {
       public CompletableFuture<Void> isBlocked();
       
       // returns null if the buffer pool is closed
       public SliceOutput take();
       
       // the call is ignored if the buffer pool is closed
       public void offer(SliceOutput buffer);
       
       // unblocks `take` and `isBlocked`
       public void close();
    }
```

Then since we want to avoid having a global lock and instead leverage a "per-partition" locking strategy we can introduce another thread safe inner class that manages writes into a single partition, e.g.:

```
    @ThreadSafe
    private static class BufferedStorageWriter
    {
        private final ExchangeStorageWriter storageWriter;
        private final BufferPool bufferPool;
        
        @GuardedBy("this")
        private SliceOutput currentBuffer;
        
        public void write(Slice slice);
       
        public CompletableFuture<Void> finish();
        
        public CompletableFuture<Void> abort();
    }
```

Then the `ExchangeSink` can use a global lock to manage a map of `BufferedStorageWriter`, e.g.:


```
@Override
    public void add(int partitionId, Slice data)
    {
        checkArgument(partitionId < outputPartitionCount, "partition id is expected to be less than %s: %s", outputPartitionCount, partitionId);

        BufferedStorageWriter writer;
        synchronized (this) {
            if (closed) {
                return;
            }
            writer = writers.computeIfAbsent(partitionId, this::createWriter);
        }
        writer.write(data);
    }
```

And have both, `finish` and `abort` protected by a global lock to simplify thinking about interactions between `finish` and `abort`